### ISSUE TITLE
Kubernetes NodePort connection only working on node running the pod, cross worker/pod connectivity not working also

### ISSUE THREAD

**Issue by acomisario** [coreos/flannel#871]
We have deployed kubernetes from scratch to understand how it works (no automation tool used like kubeadm)

## Expected Behavior
When a NodePort service is exposed on a pod (eg. TCP/31201) you should be able to get a response on the same por from any worker, also it is expected that cross pod communication works.

## Current Behavior
In this case, only the worker running the pod is responding (timeout on the rest), we understand that that's how kuberetes works

## Possible Solution
One possible solution is for documentation to be more verbose that just "kubectl create -f kube-flannel.yml" there are lots of consideration that you need to encounter through hundreds of pages based on people wit the same issue because of de documentation

BIG EXAMPLE: Do i need to deploy kube-flannel.yml but have kubelet configured with CNI? if so, what are the kubelet parameters ? how to configure CNI for flannel ? etc ...

## Steps to Reproduce (for bugs)

0. Deploy cni and their plugins on standard location /opt/cni/bin on worker nodes

1. join kubelet to the api with this parameters:
```
/usr/local/bin/kubelet \
  --hostname-override=kubernetes-node1 \
  --allow-privileged=true \
  --anonymous-auth=false \
  --authorization-mode=Webhook \
  --client-ca-file=/var/lib/kubernetes/ca.pem \
  --cluster-dns=10.32.0.10 \
  --cluster-domain=cluster.local \
  --image-pull-progress-deadline=2m \
  --kubeconfig=/var/lib/kubelet/kubeconfig \
  --pod-cidr= 10.244.0.0/16 \
  --network-plugin=cni \
  --network-plugin-dir=/etc/cni/net.d/ \
  --register-node=true \
  --require-kubeconfig \
  --runtime-request-timeout=15m \
  --tls-cert-file=/var/lib/kubelet/kubernetes-node1.pem \
  --tls-private-key-file=/var/lib/kubelet/kubernetes-node1-key.pem \
  --v=2
```
2. Deploy flannel with 
```
kubectl apply -f kube-flannel.yml
```

My kube-flannel.yml is as follow (with a little mod on cni-conf.json for the bridge to be called cbr0)
```
---
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: flannel
rules:
  - apiGroups:
      - ""
    resources:
      - pods
    verbs:
      - get
  - apiGroups:
      - ""
    resources:
      - nodes
    verbs:
      - list
      - watch
  - apiGroups:
      - ""
    resources:
      - nodes/status
    verbs:
      - patch
---
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: flannel
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: flannel
subjects:
- kind: ServiceAccount
  name: flannel
  namespace: kube-system
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: flannel
  namespace: kube-system
---
kind: ConfigMap
apiVersion: v1
metadata:
  name: kube-flannel-cfg
  namespace: kube-system
  labels:
    tier: node
    app: flannel
data:
  cni-conf.json: |
    {
      "name": "cbr0",
      "type": "flannel",
      "delegate": {
        "bridge": "cbr0",
        "hairpinMode": true,
        "isDefaultGateway": true
      }
    }
  net-conf.json: |
    {
      "Network": "10.244.0.0/16",
      "Backend": {
        "Type": "vxlan"
      }
    }
---
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  name: kube-flannel-ds
  namespace: kube-system
  labels:
    tier: node
    app: flannel
spec:
  template:
    metadata:
      labels:
        tier: node
        app: flannel
    spec:
      hostNetwork: true
      nodeSelector:
        beta.kubernetes.io/arch: amd64
      tolerations:
      - key: node-role.kubernetes.io/master
        operator: Exists
        effect: NoSchedule
      serviceAccountName: flannel
      initContainers:
      - name: install-cni
        image: quay.io/coreos/flannel:v0.9.0-amd64
        command:
        - cp
        args:
        - -f
        - /etc/kube-flannel/cni-conf.json
        - /etc/cni/net.d/10-flannel.conf
        volumeMounts:
        - name: cni
          mountPath: /etc/cni/net.d
        - name: flannel-cfg
          mountPath: /etc/kube-flannel/
      containers:
      - name: kube-flannel
        image: quay.io/coreos/flannel:v0.9.0-amd64
        command: [ "/opt/bin/flanneld", "--ip-masq", "--kube-subnet-mgr" ]
        securityContext:
          privileged: true
        env:
        - name: POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        volumeMounts:
        - name: run
          mountPath: /run
        - name: flannel-cfg
          mountPath: /etc/kube-flannel/
      volumes:
        - name: run
          hostPath:
            path: /run
        - name: cni
          hostPath:
            path: /etc/cni/net.d
        - name: flannel-cfg
          configMap:
            name: kube-flannel-cfg
```
3. Workers are ready and flannel pod is running
4. deploy a jenkins app with kubectl crate -f jenkins.yml

jenkins.yml is as follow
```
kind: Deployment
metadata:
  name: jenkins
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: jenkins
    spec:
      containers:
      - name: jenkins
        image: jenkins:2.32.2
        ports:
        - containerPort: 8080
```
5. expose the app via a nodePort service with kubectl create -f jenkis-svc.yml

jenkins-svc.yml is as follow
```
apiVersion: v1
kind: Service
metadata:
  name: jenkins
spec:
  type: NodePort
  ports:
    - port: 8080
  selector:
    app: jenkins
```
6. Check the port assigned to expose with kubectl describe svc jenkins (eg. TCP/31201)
7. Doing curl to that port from a master , only works the the ip running the pod, not all the workers respond as expected.

## Context
Im trying to run an app and expose it with kubernetes + flannel + docker
I hace already tried the 
```
iptables -P FORWARD ACCEPT
```
the only thing it does, is to allow to curl the nodePort from autside the worker itself.

## Your Environment

* Flannel version: v0.9.0-amd64
* Backend used (e.g. vxlan or udp): VXLAN
* Etcd version: 3.2.8
* Kubernetes version (if used): 1.8.0
* Operating System and version: Ubuntu 16.04
* Docker : 17.09.0-ce
* Link to your project (optional):

**Comment by acomisario** [comment]
i will close the ticket, the issue was with the VXLAN tunnels, im using AWS for the environment, and i only alowed all TCP between nodes, not UDP, so communication between workers using VXLAN was not working.

After updating the secgroup, everything is working as expected.

**Comment by vutkin** [comment]
I have same issue with 2 node cluster on virtual box sandbox.

**Comment by faicker** [comment]
@vutkin In vm environment, the tx checksum should be turned off.
ethtool -K eth0 tx off

**Comment by mbrekhov** [comment]
In my case, I configured a Kubernetes cluster on Vultr cloud. The instances on Vultr have 2 NICs - private and public-facing. 
The problem was with the Flannel Daemon set, which by default, used the public-facing NIC where UDP traffic was blocked on the firewall side.
So, I simply added '--iface=enp0s8' (with the name of my private-facing NIC) to flannel as described here: https://stackoverflow.com/a/48755233."
